### PR TITLE
Email setup: ensure hidden password reset field is validated

### PR DESCRIPTION
**Pull request by chad1008**
Related to # 8256-gh-Automattic/dotcom-forge

## Proposed Changes

- Ensure that the `PasswordResetEmailValidator` is applied even when the field is hidden, and will display the relevant field if validation fails
- Run early detection for cases where the new email domain matches the recovery email address, and expose the field when the form loads so the user can enter a valid recovery email address
- Abstract the field logic into a hook so it can be reused across the different implementations  of the field for new mailboxes

## Why are these changes being made?

Currently, no validation is run on hidden fields when setting up a new email account. This means if a user doesn't update their reset password email, and it matches the domain name they're creating the account for, they'll end up with a somewhat cryptic error at checkout, and may not know what to do. This PR updates the validation call to include hidden fields, rather than ignoring them, and sets the password reset email field visible if it fails validation. I performed an audit of the other hidden fields we track, they all look adequately typed and/or have sufficient default values that running validation on them when we weren't before won't be a problem.

After implementing the initial validation change, I really didn't like the UX of loading a mailbox creation form with a hidden error we can detect in advance, but still making the user jump through the hoop of clicking "Purchase" before the error is revealed. I experimented with running the validation on the initial render, but that results a form with an error on a prefilled field, which is terrible UX as it makes no sense for us to pre-populate what we know to be an invalid value.

In the end, I've opted for a new hook (open to feedback on the naming and file location) that runs a pseudovalidation on the domain name and user email. If a matching domain name is found, the field is not hidden, and the initial value is blank.

In effect:
- if the user has a wpcom account email that won't break our validation rules, we use it and hide the field as currently happens on `trunk`
- if the user's wpcom account email does break our validation rule, we render the field with no prefilled value, so it appears as simply another part of the signup form
- the updated validation is still in place on this PR, and can serve as a failsafe if an edge case (or perhaps a future flow not using this new hook) results in the field being hidden with an invalid value, clicking "purchase" will validate and expose the field before leading to checkout.

## Testing Instructions

Apologies for the lengthy instructions, it's largely repetition as we cover multiple flows for setting up mailboxes.

First, let's test the updated validation:

<details>
<summary>Begin by applying this test diff:</summary>

```diff
diff --git a/client/my-sites/email/hooks/use-password-reset-email-field.ts b/client/my-sites/email/hooks/use-password-reset-email-field.ts
index 205bd9fb9d..a1a55e0fb1 100644
--- a/client/my-sites/email/hooks/use-password-reset-email-field.ts
+++ b/client/my-sites/email/hooks/use-password-reset-email-field.ts
@@ -17,7 +17,7 @@ export function usePasswordResetEmailField( {
 	selectedDomainName,
 	defaultHiddenFields = [],
 }: PasswordResetEmailFieldProps ): PasswordResetEmailFieldOptions {
-	const userEmail = useSelector( getCurrentUserEmail );
+	const userEmail = 'test@YOURTESTDOMAIN.com';
 
 	// Check if the email is valid prior to official validation so we can
 	// show the field without triggering a validation error when the page
@@ -25,9 +25,7 @@ export function usePasswordResetEmailField( {
 	const isPasswordResetEmailValid = ! new RegExp( `@${ selectedDomainName }$` ).test( userEmail );
 
 	return {
-		hiddenFields: isPasswordResetEmailValid
-			? [ ...defaultHiddenFields, FIELD_PASSWORD_RESET_EMAIL ]
-			: defaultHiddenFields,
-		passwordResetEmailFieldInitialValue: isPasswordResetEmailValid ? userEmail : '',
+		hiddenFields: [ ...defaultHiddenFields, FIELD_PASSWORD_RESET_EMAIL ],
+		passwordResetEmailFieldInitialValue: userEmail,
 	};
 }

```

</details>

- Update the modified `client/my-sites/email/hooks/use-password-reset-email-field.ts` to hardcode the `userEmail` value to an address that matches your test domain (alternatively, you can go in and actually update your wpcom email account to match your test domain), but that can be a pain)
- On a site with a custom domain but no active email products, visit My Home > Upgrades > Email
- Confirm that when the form renders, the Password Reset Email field is hidden, and the textual tip with a "Chang it" link is shown
- Note: the tip will show your actual account email, not the string you just hardcoded. This is expected, the validation is going to run against your temporarily hardcoded test value.
- Enter values for the visible fields on the form
- Click "Purchase"
- Confirm the Password Reset Email field is displayed, with your mocked wpcom email account shown and the correct validation error displayed telling you a different domain is needed.
- Select the Google Workspace card on this same page and repeat the test.
- Confirm the results are the same: no field initially, and after clicking "Purchase" the field is revealed, pre-populated, and showing a validation error

Reset your local branch, but keep the hardcoded email if you'd like, we're going to test with it again

Now, lets test the updated visibility setting to avoid validation errors 

- If needed, update `client/my-sites/email/hooks/use-password-reset-email-field.ts` to hardcode the `userEmail` value to an address that matches your test domain. This will be similar to the previous diff, except it's just one change to the `userEmail` declaration within the hook. Use any address with a domain that matches your test domain.
- On a site with a custom domain but no active email products, visit My Home > Upgrades > Email.
- Confirm that on the Titan card the Password Reset Email field is visible and empty. Test the field to confirm that expected validation ( required field, make sure it's a valid email, make sure it's not the same domain) are being enforced when you click purchase.
- Test and confirm all of the same points for the Google Workspaces card on this same page

Next, we'll test the form used to set up Titan after the email onboarding flow. We'll need a new domain here.

- Update `client/my-sites/email/hooks/use-password-reset-email-field.ts` again, hardcoding `userEmail` to an address using the domain you're going to test the onboarding flow with
- Start at `/start/email-with-onboarding` and proceed through the onboarding flow, adding Professional Email along the way
- After checkout choose the option to set up your email/inbox on the thank you page
- Confirm that the password reset field renders with no prefilled value. Test to confirm the field still validates as a required field , it recognizes invalid email addresses, and it rejects email on the same domain as your new email account.
- Finish adding the initial email account

Now we can test adding additional mailboxes

- On the domain you just tested with Titan, open My Home > Upgrades > Emails and choose the option to add more mailboxes
- Assuming your userEmail is still hardcoded, confirm that the password reset field renders without a prefilled value. Test for expected validation as we've done in the previous steps

### CODE REVIEW

File: client/my-sites/email/form/mailboxes/validators.ts [hunk 1]
```diff
@@ -129,10 +130,15 @@ class MailboxNameValidator extends BaseValidator< string > {
 
 class PasswordResetEmailValidator extends BaseValidator< string > {
 	private readonly domainName: string;
+	private readonly setFieldIsVisible: MailboxForm< EmailProvider >[ 'setFieldIsVisible' ];
```
Comment (fredrikekelund) [inline client/my-sites/email/form/mailboxes/validators.ts]: The data flow in and around `NewMailBoxList` is admittedly flawed. My first instinct would have been to validate the `passwordResetEmail` field inside the `NewMailBoxList` component, but we can't do that since `hiddenFieldNames` is defined outside the component. The same limitation applies here – i.e., we can't change the field visibility from within a validator if we don't also modify `hiddenFieldNames` (and we definitely shouldn't attempt to do that from within here).

The ultimate solution would probably be to change how `hiddenFieldNames` is handled, but I'm OK with a cheaper solution for this PR. 

@chad1008, if we already have the `usePasswordResetEmailField` hook, do we really also need this `setFieldIsVisible` logic here? We already know that the default field value is OK because of the hook, and users can't edit the field without first making it visible, so an invisible invalid field seems very unlikely.

File: client/my-sites/email/form/mailboxes/validators.ts [hunk 2]
```diff
@@ -174,6 +180,7 @@ class PasswordResetEmailValidator extends BaseValidator< string > {
 			parts[ 1 ].toLowerCase() === this.domainName?.toLowerCase()
 		) {
 			field.error = PasswordResetEmailValidator.getSameDomainError( this.domainName );
+			this.setFieldIsVisible( field.fieldName, true );
```
Comment (fredrikekelund) [inline client/my-sites/email/form/mailboxes/validators.ts]:
```suggestion
			field.isVisible = true;
```
As I mentioned in my other comment, I think it's better to remove this logic. Still, theoretically, wouldn't modifying the `field.isVisible` prop directly here have the same effect?

File: client/my-sites/email/form/mailboxes/components/utilities/mailbox-operations.ts [hunk 3]
```diff
@@ -22,7 +22,7 @@ export class MailboxOperations {
 				.map( ( mailbox ) => mailbox.getFieldValue< string >( FIELD_MAILBOX ) ?? '' );
 
 			mailbox.validate(
-				true,
+				false,
```
Comment (fredrikekelund) [inline client/my-sites/email/form/mailboxes/components/utilities/mailbox-operations.ts]: Let's revert this change since the `hiddenFieldNames` prop controls which fields are hidden and the `usePasswordResetEmailField` hook unhides the password reset email field if the default value is invalid.

File: client/my-sites/email/hooks/use-password-reset-email-field.ts [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+type PasswordResetEmailFieldOptions = {
+	hiddenFields: HiddenFieldNames[];
+	passwordResetEmailFieldInitialValue: string;
+};
+
+export function usePasswordResetEmailField( {
+	selectedDomainName,
+	defaultHiddenFields = [],
+}: PasswordResetEmailFieldProps ): PasswordResetEmailFieldOptions {
+	const userEmail = useSelector( getCurrentUserEmail );
```
Comment (fredrikekelund) [inline client/my-sites/email/hooks/use-password-reset-email-field.ts:20]: Do we need to account for `userEmail` potentially being empty or null right after page load? I honestly don't know, but the thought struck me.

File: client/my-sites/email/hooks/use-password-reset-email-field.ts [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+	// Check if the email is valid prior to official validation so we can
+	// show the field without triggering a validation error when the page
+	// first loads.
+	const isPasswordResetEmailValid = ! new RegExp( `@${ selectedDomainName }$` ).test( userEmail );
+
+	return {
+		hiddenFields: isPasswordResetEmailValid
+			? [ ...defaultHiddenFields, FIELD_PASSWORD_RESET_EMAIL ]
+			: defaultHiddenFields,
+		passwordResetEmailFieldInitialValue: isPasswordResetEmailValid ? userEmail : '',
```
Comment (fredrikekelund) [inline client/my-sites/email/hooks/use-password-reset-email-field.ts]:
```suggestion
		initialValue: isPasswordResetEmailValid ? userEmail : '',
```
Optional nit. From the perspective of this hook, a shorter name like this makes more sense. It's really `MailboxesForm` etc that need the longer key name. Destructuring the key to a different variable name in the consumer is straightforward.

**Review by fredrikekelund — COMMENTED** [review]
The `usePasswordResetEmailField` approach looks pretty viable 👍 It's a shame that we can't use `PasswordResetEmailValidator` from within that hook, but replicating the validation logic in this way seems OK to me (I see we've already done so previously).

I left some comments on specific parts that I think we can revert. Let's address those, and this should hopefully be good to land after that.

File: client/my-sites/email/form/mailboxes/validators.ts (hunk 1 above)
Comment (chad1008) [reply client/my-sites/email/form/mailboxes/validators.ts]: You're right that we don't technically need it, as the hook will head off any issues. I opted to leave in this visibility logic, as well as setting `skipInvisibleFields` to `false` (one of your other comments below) to protect against either an edge case I'm not envisioning (I agree, an invalid invisible field with the new hook is very unlikely) or a potential future case where a new email setup flow is introduced, and the new hook isn't used for some reason.

It felt safer to make sure the form would validate a hidden field if the hook were ever not in play so we avoid the issue that led to this PR being necessary, but if I'm being overly cautious I'm happy to revert these parts.

> My first instinct would have been to validate the `passwordResetEmail` field inside the `NewMailBoxList` component, but we can't do that since hiddenFieldNames is defined outside the component.

Yep. I had a similar initial thought that I had to shoot down due to `hiddenFields`. Great minds...

Another advantage to doing it this way is that if we use the validator itself right off the bat, we can end up serving an error message right as the page loads, on a field we prefilled ourselves. This way we detect the error and clear the invalid value before the user sees an error they didn't even cause.

File: client/my-sites/email/form/mailboxes/validators.ts (hunk 2 above)
Comment (chad1008) [reply client/my-sites/email/form/mailboxes/validators.ts]: Will do. I liked it as a failsafe as mentioned in my other comment, but I'll remove it and we can rely on the new hook.

> Still, theoretically, wouldn't modifying the field.isVisible prop directly here have the same effect?

Yes, `field.isVisible = true` would work. I opted to use the setter out of an abundance of caution. While there isn't any logic/validation/side effects within the setter now, I thought about the possibility of something being added in the future. I didn't want a direct mutation we add today to become problematic if that property's handling ever became more complex in the future. I think that's unlikely, so it was a 'better safe than sorry' move on my part.

File: client/my-sites/email/hooks/use-password-reset-email-field.ts (hunk 4 above)
Comment (chad1008) [reply client/my-sites/email/hooks/use-password-reset-email-field.ts:20]: I think we're okay as is. It could technically be empty/null, looking at the `getCurrentUserEmail` but I think that's more likely to be if this were ever called from a logged-out state. Looking at other implementations (including on this form already, where we end up printing the the email on screen while the field is hidden) I think we're in good shape.

If we wanted to be extra cautious though, we could check the truthiness of `useEmail` while pre-checking its validity, so the form would be displayed if the current user email was missing:

```
const isPasswordResetEmailValid =
	!! userEmail && ! new RegExp( `@${ selectedDomainName }$` ).test( userEmail );
```

File: client/my-sites/email/hooks/use-password-reset-email-field.ts (hunk 5 above)
Comment (chad1008) [reply client/my-sites/email/hooks/use-password-reset-email-field.ts]: Yeah, I was actually a bit torn on the name here. I started off with exactly what you have in your suggestion, but then found myself renaming it every time I called the hook.

On the one hand I feel like if it needs to be renamed every time it's used, maybe that's a sign it isn't named well. On the other, the name within the hook only needs to make sense to the hook. Outside the hook renaming is, as you say, trivial for individual consumers to do as needed.

File: client/my-sites/email/hooks/use-password-reset-email-field.ts (hunk 4 above)
Comment (fredrikekelund) [reply client/my-sites/email/hooks/use-password-reset-email-field.ts:20]: I tested by updating the email address of my user in wp-admin, and I never ran into a case when `userEmail` was empty to start with, so I think we're fine 👍

**Review by fredrikekelund — APPROVED** [review]
LGTM and tests well 👍
